text/0000-infer-outlives.md

```diff
@@ -0,0 +1,575 @@
+
+### Long-range errors, and why they are considered unlikely
+
+Initially, we avoided inferring the `T: 'a` annotations on struct
+types in part out of a fear of "long-range" error messages, where it
+becomes hard to see the origin of an outlives requirement.  Consider
+for example a setup like this one:
+
+```rust
+struct Indirect<'a, T> {
+  field: Direct<'a, T>
+}
+
+struct Direct<'a, T> {
+  field: &'a T
+}
+```
+
+Here, both of these structs require that `T: 'a`, but the requirement
+is not written explicitly. If you have access to the full definition
+of `Direct`, it might be obvious that the requirement arises from the
+`&'a T` type, but discovering this for `Indirect` requires looking
+deeply into the definitions of all types that it references.
+
+In principle, such errors can occur, but there are many reasons to
+believe that "long-range errors" will not be a source of problems in
+practice:
+
+- The inferred bounds approach ensures that code that is given (e.g.,
+  as a parameter) an existing `Indirect` or `Direct` value will
+  already be able to assume the required outlives relationship holds.
+- Code that creates an `Indirect` or `Direct` value must also create
+  the `&'a T` reference found in `Direct`, and creating *that* reference 
+  would only be legal if `T: 'a`.
+  
+Put another way, think back on your experience writing Rust code: how
+often do you get an error that is solved by writing `where T: 'a` or
+`where 'a: 'b` **outside of a struct definition**? At least in the
+author's experience, such errors are quite infrequent.
+
+That said, long-range errors *can* still occur, typically around impls
+and associated type values, as mentioned in the previous section. For example,
+the following impl would not compile:
+
+```rust
+trait MakeRef<'a> {
+  type Type;
+}
+
+impl<'a, T> MakeRef<'a> for Vec<T> {
+  type Type = Indirect<'a, T>;
+}
+```
+
+Here, we would be missing a where-clause that `T: 'a` due to the type
+`Indirect<'a, T>`, just as we saw in the previous section. In such
+cases, tweaking the wording of the error could help to make the cause
+clearer. Similarly to auto traits, the idea would be to help trace the
+path that led to the `T: 'a` requirement on the user's behalf:
+
+```
+error[E0309]: the type `T` may not live long enough
+ --> src/main.rs:6:3
+   |
+ 6 |   type Type = Indirect<'a, T>;
+   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the type `Indirect<'a, T>` requires that `T: 'a`
+   |
+   = note: `Indirect<'a, T>` requires that `T: 'a` because it contains a field of type `Direct<'a, T>`
+   = note: `Direct<'a, T>` requires that `T: 'a` because it contains a field of type `&'a T`
+```
+
+### Impact on semver
+
+Due to the implied bounds rules, it is currently the case that
+removing `where T: 'a` annotations is potentially a breaking
+change. After this RFC, the rule is a bit more subtle: removing a
+field of type `&'a T` could affect the results of inference, and hence
```

Comment:
Adding a field could also be a breaking change. Consider this struct:
```rust
struct Foo<'a, T> {
    x: Box<Debug + 'a>,
    y: T,
}
```

Here, no `where T: 'a` bound is inferred. Consider adding a `&'a T` reference (or another type with similar WF requirements):

```rust
struct Foo<'a, T> {
    x: Box<Debug + 'a>,
    y: T,
    z: &'a T,
}
```

By adding the `z` field, we've caused `where T: 'a` to be inferred, which is a somewhat non-obvious breaking change to our type. It's especially non-obvious when you consider that `&'a T` could be replaced by `CustomRef<'a, T>`, a type which may or may not require `T: 'a`.